README
"Errno::AnyString" allows you to place an arbitrary error message in the
special $! variable, without disrupting $!'s ability to pick up the result
of the next system call that sets "errno".

It is useful if you are writing code that reports errors by setting $!, and
none of the standard system error messages fit.

  use Errno qw/EIO/;
  use Errno::AnyString qw/custom_errstr/;

  $! = custom_errstr "My hovercraft is full of eels";
  print "$!\n"; # prints My hovercraft is full of eels

  my $saved_errno = $!;

  open my $fh, "<", "/no/such/file";
  print "$!\n"; # prints No such file or directory

  $! = EIO;
  print "$!\n"; # prints Input/output error

  $! = $saved_errno;
  print "$!\n"; # prints My hovercraft is full of eels

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Errno::AnyString

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Errno::AnyString

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Errno::AnyString

    CPAN Ratings
        http://cpanratings.perl.org/d/Errno::AnyString

    Search CPAN
        http://search.cpan.org/dist/Errno::AnyString


COPYRIGHT AND LICENCE

Copyright 2009 Dave Taylor, all rights reserved.


This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Errno::AnyString

1.02    2009-08-21
        Fixed test failure under strawberry perl

1.01    2009-08-20
        Fixed failure on pre-5.7 threaded Perls (gotta love those cpan testers)
        Added a "local $!" variation to some of the quicker tests

1.00    2009-08-18
        No-longer alpha (meaning I won't lightly change the interface)
        A little more testage

0.51    2009-08-16
        Fixed failing tests under 5.10.1-RC1
        Dealt with tainted error strings.
        Auto-generate test files at packaging time, rather than build time.
        Fixed a couple of POD typos.

0.50    2009-08-14
        Added a way to permanently tie new strings to error numbers.
        Replaced the underlying mechanism with a better one.

0.05    2009-08-08
        Some POD improvements.
        Better handle the case where the string part of a saved $! value is lost.
        A few more tests.

0.04    2009-08-07
        Added documentation of the internals.
        Added an inter-operation interface.

0.03    2009-08-06
        A better interface for setting the custom error string, not backwards compatible.

0.02    2009-08-06
        Building with Makefile.PL was broken, fixed.
        More tests.

0.01    2009-08-05
        Initial release.



t/Foo1.pm
# Errno::AnyString 1.02 t/Foo1.pm
# Test module for Errno::AnyString

package Foo1;
use strict;
use warnings;

sub new {
    bless {}, shift;
}

sub errno {
    my $num = 0 + $!;
    return $num;
}

sub errstr {
    my $str = "$!";
    return $str;
}

1;



t/Foo2.pm
# Errno::AnyString 1.02 t/Foo2.pm
# Test module for Errno::AnyString

package Foo2;
use strict;
use warnings;

use English;

sub new {
    bless {}, shift;
}

sub errno {
    my $num = 0 + $OS_ERROR;
    return $num;
}

sub errstr {
    my $str = "$OS_ERROR";
    return $str;
}

1;



t/Foo3.pm
# Errno::AnyString 1.02 t/Foo3.pm
# Test module for Errno::AnyString

package Foo3;
use strict;
use warnings;

use English;

sub new {
    bless {}, shift;
}

sub errno {
    my $num = 0 + $ERRNO;
    return $num;
}

sub errstr {
    my $str = "$ERRNO";
    return $str;
}

1;



Build.PL
use 5.006;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Errno::AnyString',
    license             => 'perl',
    dist_author         => 'Dave Taylor <dave.taylor.cpan@gmail.com>',
    dist_version_from   => 'lib/Errno/AnyString.pm',
    requires => {
        'perl'         => 5.006,
        'Scalar::Util' => 0,
        'Exporter'     => 0,
        'Carp'         => 0,
    },
    build_requires => {
        'Config'             => 0,
        'English'            => 0,
        'Errno'              => 0,
        'ExtUtils::CBuilder' => 0,
        'File::Slurp'        => 0,
        'Module::Build'      => 0,
        'Scalar::Util'       => 0,
        'Test::Exception'    => 0,
        'Test::More'         => 0,
        'Test::NoWarnings'   => 0,
    },
    add_to_cleanup      => [ 'Errno::AnyString-*' ],
    create_makefile_pl => 'small',
);

$builder->create_build_script();


lib/t_compat.t.PL
use strict;
use warnings;
use File::Slurp;

my $text = join '', <DATA>;

$_ = $text;
s/---localise---//;
write_file "t/compat-nolocal.t", $_;

$_ = $text;
s/---localise---/local \$!;/;
write_file "t/compat-local.t", $_;

__DATA__
# Automatically generated by Errno::AnyString 0.50 lib/t_compat.t.PL
# Test compatibility with traditional $! behavior.

use strict;
use warnings;

use Test::More;
use Test::NoWarnings;

use Scalar::Util qw/dualvar/;
use Errno ':POSIX';

use lib 't';
use NormalErrnoOperation;

our @system_errors = NormalErrnoOperation->new;
our $fh_none = NormalErrnoOperation->flathash_none;

my $tests_per_errno_ok = 6 + @system_errors;
my $errno_ok_calls = 2 * 2 * (3 * @system_errors * 9);
plan tests => $errno_ok_calls * $tests_per_errno_ok + 3 + 2*2*8 + 1;

our $errno_ok_start_with_numeric;

foreach my $testtype (qw/baseline compat compat2/) {
    if ($testtype eq 'compat') {
        eval 'use Errno::AnyString qw/custom_errstr register_errstr/'; die $@ if $@;
    }
    ---localise---
    $! = 74;
    my $copied_errno = $!;
    foreach my $reset_zero (0, 1) {
        foreach $errno_ok_start_with_numeric (0, 1) {
            foreach my $i (0 .. $#system_errors) {
                my $sym = $system_errors[$i]->symbol; # Errno.pm symbolic code, such as ENOENT
                my $errno = $system_errors[$i]->errno;
                my $name = $sym || $errno;

                $system_errors[$i]->set;
                errno_ok( $system_errors[$i], "$testtype setter $name" );
                $! = 0 if $reset_zero;

                {
                    local $!;
                    $system_errors[$i]->set;
                    errno_ok( $system_errors[$i], "$testtype setter $name, local preserves magic" );
                }

                $! = $errno;
                errno_ok( $system_errors[$i], "$testtype set errno $name" );
                $! = 0 if $reset_zero;

                # anything that perl groks as a number must be treated as such
                $! = $errno."e0";
                errno_ok( $system_errors[$i], "$testtype set errno $name exp" );
                $! = 0 if $reset_zero;

                $! = "$errno ";
                errno_ok( $system_errors[$i], "$testtype set errno $name trailspace" );
                $! = 0 if $reset_zero;

                $! = "$errno.0000e00 ";
                errno_ok( $system_errors[$i], "$testtype set errno $name xmas" );
                $! = 0 if $reset_zero;

                # even this silly thing must work. It warns, but sets $! to the number
                { 
                    no warnings;
                    $! = "$errno things are currently on fire";
                }
                errno_ok( $system_errors[$i], "$testtype set errno $name silly" );
                $! = 0 if $reset_zero;

                # Native $! will ignore the pv if the sv contains a number, so do I
                # unless my magic errno value is used.
                $! = dualvar $errno, "123456";
                errno_ok( $system_errors[$i], "$testtype set errno $name dualvar" );
                $! = 0 if $reset_zero;

                if ($sym) {
                    eval "\$! = $sym"; die $@ if $@;
                    errno_ok( $system_errors[$i], "$testtype set symbol $name" );
                    $! = 0 if $reset_zero;
                } else {
                    foreach my $j (1 .. $tests_per_errno_ok) {
                        ok 1, "fake test to simplify test counting calculation";
                    }
                }
            }
            if ($testtype eq "compat") {
                $! = custom_errstr("my hovercraft is full of eels");
                my $fh_now = NormalErrnoOperation->flathash_now;
                is $fh_now, $fh_none, "custom errstr sets nothing in %!";
                is $!+0, 458513437, "custom string expected errno";
                is "$!", "my hovercraft is full of eels", "custom string expected errstr";
                $! = 0 if $reset_zero;
                
                $! = register_errstr("my hovercraft is full of peels");
                $fh_now = NormalErrnoOperation->flathash_now;
                is $fh_now, $fh_none, "register errstr sets nothing in %!";
                is "$!", "my hovercraft is full of peels", "register string expected errstr";
                $! = 0 if $reset_zero;

                $! = register_errstr("my hovercraft is full of beagles", 654321);
                $fh_now = NormalErrnoOperation->flathash_now;
                is $fh_now, $fh_none, "register errstr 654321 sets nothing in %!";
                is $!+0, 654321, "custom string expected errno";
                is "$!", "my hovercraft is full of beagles", "register string 654321 expected errstr";
                $! = 0 if $reset_zero;
            }
        }
    }
    is 0+$copied_errno, 74, "copying $! does not copy the magic";
}

sub errno_ok {
    my ($want, $testname) = @_;

    $testname .= " $errno_ok_start_with_numeric";

    {
        local $! = 13;
        is 0+$!, 13, "$testname localized 13";
    }

    if ($errno_ok_start_with_numeric) {
        is 0+$!, $want->errno, "$testname errno";
    }
    is "$!", $want->errstr, "$testname errstr";
    is 0+$!, $want->errno, "$testname errno 2";
    is "$!", $want->errstr, "$testname errstr 2";
    unless ($errno_ok_start_with_numeric) {
        is 0+$!, $want->errno, "$testname errno postponed";
    }
    foreach my $other (@system_errors) {
        if ($other->errno == $want->errno and $other->symbol) {
            ok $!{$other->symbol}, "$testname correct symbol set in %!";
        } elsif ($other->symbol) {
            ok ! $!{$other->symbol}, "$testname incorrect symbol not set in %!";
        } else {
            ok 1, "fake test to simplify test counting calculation";
        }
    }
    is $want->flathash_now, $want->normal_flathash, "$testname %!";
}



lib/t_autodie.t.PL
use strict;
use warnings;
use File::Slurp;

my $text = join '', <DATA>;

# Shall I test against autodie.pm or Fatal.pm ?
my @mods = qw/autodie Fatal/;

# Shall I load autodie/Fatal before or after Errno::AnyString ?
my %how = (
    before => '',
    after  => 'BEGIN { require Errno::AnyString; }',
);

foreach my $mod (@mods) {
    my $stage1 = $text;
    $stage1 =~ s/---mod---/$mod/g or die;
    while ( my ($name, $how) = each %how ) {
        $_ = $stage1;
        s/---name---/$name/ or die;
        s/---load-me---/$how/ or die;
        write_file "t/with-$mod-$name.t", $_;
    }
}

__DATA__
# Automatically generated by Errno::AnyString 0.50 lib/t_autodie.t.PL
# Inter-operation with autodie, ---mod--- loaded ---name--- Errno::AnyString

use strict;
use warnings;

---load-me---

use Test::More;
use Test::Exception;
BEGIN {
    eval 'require ---mod---';
    plan skip_all => '---mod--- required' if $@;

    plan tests => 3;
}
use Test::NoWarnings;

sub try_but_fail1 {
    $! = custom_errstr("set with custom errstr");
    return;
}
sub try_but_fail2 {
    $! = 999_999_999;
    return;
}
use ---mod--- qw/try_but_fail1 try_but_fail2/;

use Errno::AnyString qw/custom_errstr register_errstr/;
register_errstr "set with register errstr", 999_999_999;

throws_ok { try_but_fail1() } qr/set with custom errstr/,   "custom errstr seen by ---mod---";
throws_ok { try_but_fail2() } qr/set with register errstr/, "register errstr seen by ---mod---";



lib/in-pm-files.t.PL
# Errno::AnyString 0.01 lib/in-pm-files.t.PL
# Check that the effects of loading Errno::AnyString materialise in other
# already-loaded modules.

use strict;
use warnings;
use File::Slurp;

my %use_errno = (
    none => '# none',
    use  => 'use Errno;',
    euse => 'eval "use Errno"; die $@ if $@;',
    impl => 'my @implicitly_load_Errno_pm = keys(%!);',
);
my $data = join '', <DATA>;

my $mod_count = 4;
my $pkg_names = join " ", map {"Foo$_"} (1 .. $mod_count);
my $pkg_use = join "", map {"use Foo$_;\n"} (1 .. $mod_count);

foreach my $use (keys %use_errno) {
    $_ = $data;
    s/---use_errno---/$use_errno{$use}/g;
    s/---pkg_use---/$pkg_use/g;
    s/---pkg_names---/$pkg_names/g;
    write_file "t/in-pm-files-$use.t", $_;
}

__DATA__
# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
#
# This file was auto-generated by lib/in-pm-files.t.PL
#
# Make changes there, not here.
#
# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

use strict;
use warnings;

# The line below differs between t/in-pm-files-*.t - it pulls
# in Errno.pm (or not) in various different ways.
---use_errno---

use Test::More;
use Test::NoWarnings;

use lib 't';
---pkg_use---
use Errno::AnyString qw/ERRSTR_SET/;

my @pkgs = qw/---pkg_names---/;
plan tests => @pkgs*6 + 1;

foreach my $pkg (@pkgs) {
    my $x = $pkg->new;

    $! = "--string1";
    is $x->errstr, "string1", "$pkg special errstr set";
    is $x->errno, ERRSTR_SET, "$pkg special errno set";

    my $save = $!;

    $! = 0;
    isnt $x->errstr, "string1", "$pkg special errstr unset";
    isnt $x->errno, ERRSTR_SET, "$pkg special errno unset";

    $! = $save;
    is $x->errstr, "string1", "$pkg special errstr restored";
    is $x->errno, ERRSTR_SET, "$pkg special errno restored";
}



lib/t_in-pm-files.t.PL
use strict;
use warnings;
use File::Slurp;

my $text = join '', <DATA>;

# Different ways to load Errno (or not)
my %how = (
    none     => '',
    use      => 'use Errno;',
    evaluse  => 'eval "use Errno"; die $@ if $@;',
    implicit => 'my @implicitly_load_Errno_pm = keys(%!);',
);

while ( my ($name, $how) = each %how ) {
    $_ = $text;
    s/---name---/$how/;
    s/---UseErrno---/$how/ or die;
    write_file "t/in-pm-files-$name.t", $_;
}

__DATA__
# Automatically generated by Errno::AnyString 0.50 lib/t_in-pm-files.t.PL
# Check that the effects of loading Errno::AnyString materialize in other
# already-loaded modules.  Errno.pm loading = ---name---

use strict;
use warnings;

---UseErrno---

use Test::More;
use Test::NoWarnings;

use lib 't';
use Foo1;
use Foo2;
use Foo3;
use Foo4;
use Errno::AnyString qw/CUSTOM_ERRSTR_ERRNO custom_errstr/;

my @pkgs = map {"Foo$_"} (1 .. 4);
plan tests => @pkgs*6 + 1;

foreach my $pkg (@pkgs) {
    my $x = $pkg->new;

    $! = custom_errstr "string1";
    is $x->errstr, "string1", "$pkg special errstr set";
    is $x->errno, CUSTOM_ERRSTR_ERRNO, "$pkg special errno set";

    my $save = $!;

    $! = 0;
    isnt $x->errstr, "string1", "$pkg special errstr unset";
    isnt $x->errno, CUSTOM_ERRSTR_ERRNO, "$pkg special errno unset";

    $! = $save;
    is $x->errstr, "string1", "$pkg special errstr restored";
    is $x->errno, CUSTOM_ERRSTR_ERRNO, "$pkg special errno restored";
}



lib/Errno/AnyString.pm
package Errno::AnyString;
use strict;
use warnings;

=head1 NAME

Errno::AnyString - put arbitrary strings in $!

=head1 VERSION

Version 1.02

=cut

our $VERSION = '1.02';

=head1 SYNOPSIS

C<Errno::AnyString> allows you to place an arbitrary error message in the special C<$!> variable, without disrupting C<$!>'s ability to pick up the result of the next system call that sets C<errno>.

It is useful if you are writing code that reports errors by setting C<$!>, and none of the standard system error messages fit.

  use Errno qw/EIO/;
  use Errno::AnyString qw/custom_errstr/;

  $! = custom_errstr "My hovercraft is full of eels";
  print "$!\n"; # prints My hovercraft is full of eels

  my $saved_errno = $!;

  open my $fh, "<", "/no/such/file";
  print "$!\n"; # prints No such file or directory

  $! = EIO;
  print "$!\n"; # prints Input/output error

  $! = $saved_errno;
  print "$!\n"; # prints My hovercraft is full of eels

You can also set the error strings for particular error numbers, for the lifetime of the Perl interpreter:

  use Errno::AnyString qw/register_errstr/;

  register_errstr "Wetware failure", 339864;

  $! = 339864;
  print "$!\n"; # prints Wetware failure



=head1 BACKGROUND

Perl's special C<$!> variable provides access to  C<errno>, the "error number", which is an integer variable used by C library functions to record what went wrong when they fail. See L<perlvar/ERRNO> and L<Errno>.

There is a fixed error message for each C<errno> value in use, and a C library function to translate C<errno> values into error messages. The magical C<$!> variable always holds the current value of C<errno> if you use it in a numeric context, and the corresponding error message if you use it in a string context.

  open my $fh, "<", "/no/such/file"; # the failing open sets errno to 2
  my $errno = $! + 0;  # $errno now contains 2
  my $err = "$!";      # $err now contains "No such file or directory"

You can also assign a number to C<$!>, to set the value of C<errno>. An C<errno> value of 22 means "invalid argument", so:

  $! = 22;
  $errno = $! + 0;  # $errno now contains 22
  $err = "$!";      # $err now contains "Invalid argument"

What you can't do however is assign a string of your own choice to C<$!>. If you try, Perl just converts your string to an integer as best it can and puts that in C<errno>.

  $! = "You broke it";
  # gives an "Argument isn't numeric" warning and sets errno to 0

=head1 DESCRIPTION

C<Errno::AnyString> allows you to set the error message strings that correspond to particular C<errno> values. It makes a change to the C<$!> magic so that the correct string is returned when C<errno> takes a value for which a string has been registered. The change to C<$!> is global and lasts until the Perl interpreter exits.

=cut

use Exporter;
use Carp;
use Scalar::Util qw/dualvar tainted/;

require XSLoader;
XSLoader::load('Errno::AnyString', $VERSION);

our @ISA = qw/Exporter/;
our @EXPORT_OK = qw/custom_errstr register_errstr CUSTOM_ERRSTR_ERRNO/;

our (%Errno2Errstr, %_string2errno);

Errno::AnyString::_install_my_magic($!);

=head1 EXPORTS

Nothing is exported by default. The following are available for export.

=head2 CUSTOM_ERRSTR_ERRNO

A constant with the value 458513437. This is the C<errno> value used by this module to indicate that a custom error string set with custom_errstr() is active. This value was chosen at random, to avoid picking an C<errno> value that something else uses.

=cut

sub CUSTOM_ERRSTR_ERRNO () { return 458513437; }

our $_next_registered_errno = CUSTOM_ERRSTR_ERRNO() + 1;

=head2 custom_errstr ( ERROR_STRING )

Returns a value which will set the specified custom error string when assigned to C<$!>. 

The returned value is actually a dual valued scalar with C<CUSTOM_ERRSTR_ERRNO> as its numeric value and the specified error string as its string value. It's not just magical variables like C<$!> that can hold a number and a string at the same time, ordinary Perl scalars can do it as well. See L<Scalar::Util/dualvar>.

With C<Errno::AnyString> loaded, the C<$!> magic responds specially to a scalar with a numeric value of CUSTOM_ERRSTR_ERRNO being assigned to C<$!>: the string value of the scalar gets recorded as the registered string for C<errno> value CUSTOM_ERRSTR_ERRNO, replacing any previous registered string for that value.

This way of setting the custom error string was chosen because it works well with code that saves and restores the value of C<$!>.

  $! = custom_errstr "Test string";

  my $saved_errno = $!;
  do_other_things();
  $! = $saved_errno;

  print "$!\n"; # prints Test string

When C<$!> is copied to C<$saved_errno>, C<$saved_errno> becomes dual valued with a number value of CUSTOM_ERRSTR_ERRNO and a string value of "Test string". When C<$saved_errno> gets copied back to C<$!> at the end, the number value of CUSTOM_ERRSTR_ERRNO triggers the modified C<$!> magic to register the string value of "Test string" as the custom error string for CUSTOM_ERRSTR_ERRNO.

This is important because code called from within do_other_things() might itself use custom_errstr() to set custom error strings, overwriting the registered error string of "Test string". Since C<$saved_errno> saves the error message string as well as the C<errno> value, the C<$!> magic can put the correct string back in place when the C<$saved_errno> value is restored.

=cut

sub custom_errstr ($) {
    tainted $_[0] and croak "Tainted error string used with Errno::AnyString";

    return dualvar CUSTOM_ERRSTR_ERRNO, "$_[0]";
}

=head2 register_errstr ( ERROR_STRING [,ERRNO_VALUE] )

register_errstr() can be used in a similar way to custom_errstr():

  $! = register_errstr "An error string";

The difference is that register_errstr() permanently (i.e. for the lifetime of the Perl interpreter) assigns an C<errno> value to that error string. The error string is stored away, and will be used as the string value of C<$!> any time that C<errno> value is set in future. By default, register_errstr() picks a large C<errno> value that it has not yet assigned to any other string. 

If you call register_errstr() repeatedly with the same error string, it will notice and use the same C<errno> value each time. That means it's safe to do something like:

  $! = register_errstr "Too many foos defined";

in code that could be called a large number of times, and register_errstr() will store only one copy of the string and use up only one C<errno> value.

You can specify the C<errno> value to use as a second parameter to register_errstr(), for example:

  $! = register_errstr "my error string", 999999;

This sets the error string for C<errno> value 999999 (replacing any previously set error string for 999999) and assigns it to C<$!>. You can also call register_errstr() simply to register a bunch of new error codes, without assigning the return value to C<$!> each time:

  register_errstr "Invalid foodb file",      -20000;
  register_errstr "Invalid foodb parameter", -20001;
  register_errstr "foodb out of key slots",  -20002;

  # ...

  $! = -20001;
  print "$!\n"; # prints Invalid foodb parameter

It is also possible to use register_errstr() to replace the standard system error messages. For example, to replace the "Permission denied" message;

  use Errno qw/EACCES/;
  register_errstr "I'm sorry, Dave. I'm afraid I can't do that", EACCES;

  open my $fh, ">/no_permission_to_write_here";
  print "$!\n"; prints "I'm sorry, Dave. I'm afraid I can't do that"

This is not something I'd recommend, as it's likely to cause confusion. In general, when specifying the C<errno> value to register_errstr() one should take care to avoid values that are likely to be used for any other purpose.

Internally, the error strings registered by register_errstr() are kept in the C<%Errno::AnyString::Errno2Errstr> hash. You shouldn't go poking around in this hash yourself, but by localising it you can limit the scope of a register_errstr() registration:
  
  {
      local %Errno::AnyString::Errno2Errstr = %Errno::AnyString::Errno2Errstr;

      register_errstr "I'm sorry, Dave. I'm afraid I can't do that", EACCES;

      # here you have a silly error message in place of "Permission denied"
  }
  # here sanity is restored

=cut

sub register_errstr ($;$) {
    my ($str, $num) = @_;

    tainted $str and croak "Tainted error string used with Errno::AnyString";

    unless (defined $num) {
        $num = $_string2errno{$str};
        unless (defined $num) {
            $num = $_next_registered_errno++;
            $_string2errno{$str} = $num;
        }
    }
    $Errno2Errstr{$num} = "$str";

    return dualvar $num, $str;
}

=head1 INTER-OPERATION

This section is aimed at the authors of other modules that alter C<$!>'s behaviour, as a guide to ensuring clean inter-operation between Errno::AnyString and your module.

Errno::AnyString works by adding two instances of uvar magic to C<$!>, one at the head of the list and one at the tail. It does not modify or remove any existing magic from C<$!>. It should inter-operate cleanly with anything else that adds more magic to C<$!>, so long magic is added in a way that preserves existing uvar magic.

Emptying the C<%Errno::AnyString::Errno2Errstr> hash effectively turns off this module's interference with C<$!>, so you can get a "real" C<$!> value with:

  my $e = do { local %Errno::AnyString::Errno2Errstr ; $! };

=head1 AUTHOR

Dave Taylor, C<< <dave.taylor.cpan at gmail.com> >>

=head1 BUGS AND LIMITATIONS

=head2 C LEVEL STRERROR CALLS

If C level code attempts to get a textual error message based on C<errno> while a custom error string is set, it will get something like the following, depending on the platform:

  Unknown error 458513437

=head2 PURE NUMERIC RESTORE

If the string part of a saved custom_errstr() C<$!> value is lost, then restoring that value to C<$!> restores the string most recently set with custom_errstr(), which is not necessarily the string that was set when the C<$!> value was saved.

  $! = custom_errstr "String 1";
  my $saved_errno = 0 + $!;

  $! = custom_errstr "String 2";

  $! = $saved_errno;
  print "$!\n"; # prints String 2

Note that the Perl code that saved the error number had to go out of its way to discard the string part of C<$!>, so I think this combination is fairly unlikely in practice.

Error strings set with register_errstr() are not effected by this issue, since each gets its own unique C<errno> value. For this reason, register_errstr() should be used in preference to custom_errstr() if you have a small number of fixed error strings:

  $! = register_errstr "Attempt to frob without a foo"; # good

However, register_errstr() uses up an C<errno> value and permanently stores the string each time it is called with a string it has not seen before. If your code could generate a large number of different error strings over the lifetime of the Perl interpreter, then using register_errstr() could cost a lot of memory. In such cases, custom_errstr() would be a better choice.

  $! = register_errstr "failed at $line in $file: $why"; # less good 

=head2 TAINT MODE

I'm currently unable to find a good way to propagate the taintedness of custom error strings through C<$!>, due to an interaction between taint magic, C<$!>'s dualvar behaviour and the SvPOK flag in some Perl versions. If Perl is in taint mode then passing a tainted error string to custom_errstr() or register_errstr() will cause an immediate croak with the message:

  Tainted error string used with Errno::AnyString

=head2 OTHER BUGS

Please report any other bugs or feature requests to C<bug-errno-anystring at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Errno::AnyString>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Errno::AnyString

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Errno::AnyString>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Errno::AnyString>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Errno::AnyString>

=item * Search CPAN

L<http://search.cpan.org/dist/Errno::AnyString>

=back

=head1 SEE ALSO

L<Errno>, L<perlvar/ERRNO>, L<Scalar::Util>, L<perlguts>

=head1 COPYRIGHT & LICENSE

Copyright 2009 Dave Taylor, all rights reserved.


This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Errno::AnyString
